crates/nu-system/src/foreground.rs

```diff
@@ -0,0 +1,111 @@
+use std::process::{Child, Command};
+
+/// A simple wrapper for `std::process::Command`
+///
+/// ## spawn behavior
+/// ### Unix
+/// When invoke `spawn`, current process will ignore `SIGTTOU`, `SIGTTIN` signal, spawned child process will get it's
+/// own process group id, and it's going to foreground(by making stdin belong's to child's process group).
+///
+/// When child is to over, `SIGTTOU`, `SIGTTIN` signal will be reset, foreground process is back to callers' process.
+///
+/// ### Windows
+/// It does nothing special on windows system, `spawn` is the same as [std::process::Command::spawn](std::process::Command::spawn)
+pub struct ForegroundProcess {
+    inner: Command,
+}
+
+/// A simple wrapper for `std::process::Child`
+///
+/// It can only be created by `ForegroundProcess::spawn`.
+pub struct ForegroundChild {
+    inner: Child,
+}
+
+impl ForegroundProcess {
+    pub fn new(cmd: Command) -> Self {
+        Self { inner: cmd }
+    }
+
+    pub fn spawn(&mut self) -> std::io::Result<ForegroundChild> {
+        fg_process_setup::prepare_to_foreground(&mut self.inner);
+        self.inner.spawn().map(|child| {
+            fg_process_setup::set_foreground(&child);
+            ForegroundChild { inner: child }
+        })
+    }
+}
+
+impl AsMut<Child> for ForegroundChild {
+    fn as_mut(&mut self) -> &mut Child {
+        &mut self.inner
+    }
+}
+
+impl Drop for ForegroundChild {
+    fn drop(&mut self) {
+        // It's ok to use here because we have called `set_foreground` during creation.
+        unsafe { fg_process_setup::reset_foreground_id() }
+    }
+}
+
+// It's a simpler version of fish shell's external process handling.
+//
+// For more information, please check `child_setup_process` function in fish shell.
+// https://github.com/fish-shell/fish-shell/blob/3f90efca38079922b4b21707001d7bb9630107eb/src/postfork.cpp#L140
+#[cfg(target_family = "unix")]
+mod fg_process_setup {
+    use std::os::unix::prelude::CommandExt;
+    pub fn prepare_to_foreground(external_command: &mut std::process::Command) {
```

Comment:
With the wrapping types, the helpers probably don't have to be `pub`

Reply:
I've tried remove `pub`, the compiler gives me error says that they are private

Reply:
Right they are in their own module, thus requiring a visibility modifier to be visible above in the module hierarchy. 
It is probably a stylistic question. I would prefer doing the `cfg()` switch inside of the functions (one definition to find, that makes the platform differences an obvious implementation detail). This could either omit the module entirely, else marking it as `pub(super)` should denote that it should only be visible in the parent module.

Reply:
Ok, I think I get your point, so I remove the extra inner module